File: proposed-projects/round-trip-experiment/reverse/20260909T151143Z_170b737d/20260910T085124Z_7f672549/ported_implementation/gbnf/grammar_graph/pointers.py
from typing import Generator

from .graph_pointer import GraphPointer


class Pointers:
    def __init__(self, *pointers: GraphPointer):
        self.__pointers__ = {}
        for pointer in pointers:
            self.add(pointer)

    def __repr__(self) -> str:
        return f"<Pointers {', '.join(str(p.node.rule) for p in self)}>"

    def add(self, pointer: GraphPointer) -> None:
        self.__pointers__[pointer.id] = pointer

    def __iter__(self) -> Generator[GraphPointer, None, None]:
        yield from list(self.__pointers__.values())

    def __len__(self) -> int:
        return len(self.__pointers__)

File: packages/gbnf-experiment/data/20260909T144331Z_5b46f630/ported_implementation/src/grammarGraph/pointers.ts
import type { GraphPointer, GraphPointerKey } from './graphPointer.js';

export class Pointers {
  #pointers: Map<GraphPointerKey, GraphPointer>;

  constructor(...pointers: GraphPointer[]) {
    this.#pointers = new Map();
    for (const pointer of pointers) {
      this.add(pointer);
    }
  }

  add(pointer: GraphPointer): void {
    this.#pointers.set(pointer.id, pointer);
  }

  *[Symbol.iterator](): Generator<GraphPointer> {
    yield* this.#pointers.values();
  }

  get size(): number {
    return this.#pointers.size;
  }

  toString(): string {
    return `<Pointers ${[...this].map((p) => p.node.rule).join(', ')}>`;
  }
}
